[SYCL][UR] Link UR PI against UR Loader

bmyates opened the pull request:
Compile UR L0 adapter as a standalone library

Signed-off-by: Brandon Yates <brandon.yates@intel.com>

smaslov-intel commented on sycl/plugins/unified_runtime/CMakeLists.txt:
```diff
@@ -63,20 +58,39 @@ add_sycl_plugin(unified_runtime
+)
+
+# Build level zero adapter
+add_sycl_library("ur_adapter_level_zero" SHARED
```
what will cause the UR loader to discover and load this adapter?

bmyates replied:
There are couple additional steps to make this adapter loadable.  

1) Register adapter library filename with loader.  I opened PR here: https://github.com/oneapi-src/unified-runtime/pull/353/commits/53e7e7c1a4bc56e43411807602c7fc5cf4c6c18a
2) Add implementations of urGet*ProcAddrTable functions into adapter. 

Until both are completed sycl-ls will not work correctly.    I opened this PR as-is because there was some mention that this was blocking cuda adapter development. 

I will start working on adding (2) and validating with sycl-ls.

bmyates replied:
My latest updated is fully loadable and sycl-ls works as expected

jandres742 replied:
confirmed on my side.

igchor commented on sycl/plugins/unified_runtime/CMakeLists.txt:
```diff
@@ -63,20 +58,39 @@ add_sycl_plugin(unified_runtime
+  INCLUDE_DIRS
+  LIBRARIES
     UnifiedRuntime-Headers
+    UnifiedRuntimeLoader
```
Why do you need to link with the loader here?

bmyates replied:
good catch I don't think it is needed, i will remove

igchor commented on sycl/plugins/unified_runtime/CMakeLists.txt:
```diff
@@ -63,20 +58,39 @@ add_sycl_plugin(unified_runtime
     "ur/ur.cpp"
-    "ur/adapters/level_zero/ur_level_zero.hpp"
-    "ur/adapters/level_zero/ur_level_zero.cpp"
     "ur/usm_allocator.hpp"
```
I think ur.*pp and usm_allocator.*pp files should not be needed here. They should be used to build ur_adapter_level_zero

bmyates replied:
thanks will fix

smaslov-intel commented on sycl/plugins/unified_runtime/pi2ur.hpp:
```diff
@@ -242,6 +241,7 @@ namespace pi2ur {
 inline pi_result piPlatformsGet(pi_uint32 num_entries, pi_platform *platforms,
                                 pi_uint32 *num_platforms) {
 
+  urInit(0);
```
I think this should go inside `urPlatformGet`

bmyates replied:
urInit has to be called before any other ur api

smaslov-intel replied:
then that should be called from SYCL RT directly, no?

bmyates replied:
Do you have a suggestion for where to put it? Some initialization path somewhere?

jandres742 replied:
@smaslov-intel : I think this place is correct. piPlatformsGet is typically the first function to be called, and then this calls urPlatformsGet, which then calls zeInit in the L0 adapter.

smaslov-intel replied:
I think it is OK for now to keep it here, but when we move away from PI (SYCL RT calls UR directly) it should be explicitly called from SYCL RT.

smaslov-intel commented on sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero.hpp:
```diff
@@ -107,8 +107,6 @@ static ur_result_t ze2urResult(ze_result_t ZeResult) {
       {ZE_RESULT_ERROR_OVERLAPPING_REGIONS, UR_RESULT_ERROR_INVALID_OPERATION},
       {ZE_RESULT_ERROR_INVALID_GROUP_SIZE_DIMENSION,
        UR_RESULT_ERROR_INVALID_WORK_GROUP_SIZE},
-      {ZE_RESULT_ERROR_MODULE_BUILD_FAILURE,
```
why did you remove this?

bmyates replied:
`UR_RESULT_ERROR_MODULE_BUILD_FAILURE`  was removed from UR spec.  I do not know why and I don't see an obvious replacement

https://github.com/oneapi-src/unified-runtime/commit/d7e4d2b67c56123b37fe7b3d558a70ba853cd311

jandres742 replied:
@smaslov-intel : change from @bmyates  is correct. Once this merged, I will rebased https://github.com/intel/llvm/pull/8402 with latest changes in interfaces.

smaslov-intel replied:
@jandres742 : but you will need this error in UR for here: https://github.com/intel/llvm/blob/30c80625ad339aadbb49156c2cc3eef140d7b0dd/sycl/plugins/level_zero/pi_level_zero.cpp#LL3713C2-L3713C2

jandres742 replied:
@smaslov-intel : ah I see what you mean. You mean we cannot remove the entry but replace it. See below to what I actually did while testing @bmyates  patch locally:

```cpp
     {ZE_RESULT_ERROR_MODULE_BUILD_FAILURE,
-     UR_RESULT_ERROR_MODULE_BUILD_FAILURE},
+     UR_RESULT_ERROR_PROGRAM_BUILD_FAILURE},
```

is that what you mean?

smaslov-intel replied:
> @smaslov-intel : ah I see what you mean. You mean we cannot remove the entry but replace it. See below to what I actually did while testing @bmyates patch locally:
> 
> ```c++
>      {ZE_RESULT_ERROR_MODULE_BUILD_FAILURE,
> -     UR_RESULT_ERROR_MODULE_BUILD_FAILURE},
> +     UR_RESULT_ERROR_PROGRAM_BUILD_FAILURE},
> ```
> 
> is that what you mean?

Yep, looks like they just renamed module to program, so we need to replace.

smaslov-intel commented on sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_loader_interface.cpp:
How is this new file used?

bmyates replied:
The UR loader calls these functions to build the DDI tables.

jandres742 replied:
@bmyates  this is the extra level of indirection that we dont want to have when there's only one adapter, right? that optimization could be done in a  follow-up patch,.

jandres742 reviewed (commented):
+1

smaslov-intel reviewed (approved):
LGTM

smaslov-intel commented on sycl/plugins/unified_runtime/CMakeLists.txt:
```diff
@@ -17,6 +17,7 @@ if (NOT DEFINED UNIFIED_RUNTIME_LIBRARY OR NOT DEFINED UNIFIED_RUNTIME_INCLUDE_D
   set(CMAKE_CXX_FLAGS_BAK "${CMAKE_CXX_FLAGS}")
   if (UNIX)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-error")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-suggest-override")
```
Why -Wno-error above is not enough? It was meant to ignore all warnings when built from SYCL

smaslov-intel replied:
See https://github.com/intel/llvm/pull/8683 where we are trying to fix UR build

bmyates replied:
I had a build failure I thought was caused by this warning in loader.  Checking again, this isn't needed. 

I removed it and rebased this on #8683

jandres742 commented:
@bader : do you know what could be happening with the CI here?

bader commented:
> @bader : do you know what could be happening with the CI here?

CI works perfectly fine. This is a bug in CMake files. It should be fixed by https://github.com/intel/llvm/commit/c39de7135bcd1629fa09a0f15a24ab7c23a85516. Please, update your branch.

bmyates commented:
> > @bader : do you know what could be happening with the CI here?
> 
> CI works perfectly fine. This is a bug in CMake files. It should be fixed by [c39de71](https://github.com/intel/llvm/commit/c39de7135bcd1629fa09a0f15a24ab7c23a85516). Please, update your branch.

I rebased and fixed that issue, but now failing with

bader commented:
> > > @bader : do you know what could be happening with the CI here?
> > 
> > 
> > CI works perfectly fine. This is a bug in CMake files. It should be fixed by [c39de71](https://github.com/intel/llvm/commit/c39de7135bcd1629fa09a0f15a24ab7c23a85516). Please, update your branch.
> 
> I rebased and fixed that issue, but now failing with another issue that seems unrelated. Any idea @bader `/usr/bin/ld: /tmp/lit-tmp-8505fbty/level_zero_imm_cmdlist_per_thread-d70b3a.o: undefined reference to symbol 'pthread_create@@GLIBC_2.2.5`

No idea.

bmyates commented:
> > > @bader : do you know what could be happening with the CI here?
> > 
> > 
> > CI works perfectly fine. This is a bug in CMake files. It should be fixed by [c39de71](https://github.com/intel/llvm/commit/c39de7135bcd1629fa09a0f15a24ab7c23a85516). Please, update your branch.
> 
> I rebased and fixed that issue, but now failing with another issue that seems unrelated. Any idea @bader `/usr/bin/ld: /tmp/lit-tmp-8505fbty/level_zero_imm_cmdlist_per_thread-d70b3a.o: undefined reference to symbol 'pthread_create@@GLIBC_2.2.5`

This was caused by a bug in sycl tests that was fixed yesterday https://github.com/intel/llvm-test-suite/pull/1674

Retriggered CI here, fingers crossed it passes this time.

bader commented on sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero.cpp:
```diff
@@ -1720,3 +1720,9 @@ UR_APIEXPORT ur_result_t UR_APICALL urDevicePartition(
+  return UR_RESULT_SUCCESS;
+}
+
+ur_result_t urTearDown(void *pParams) { return UR_RESULT_SUCCESS; }
```
```suggestion
ur_result_t urTearDown(void *pParams) { return UR_RESULT_SUCCESS; }

```

bader commented on sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_loader_interface.cpp:
```diff
@@ -0,0 +1,291 @@
+//===--------- ur_loader_interface.cpp - Level Zero Adapter
+//-----------------===//
```
```suggestion
//===--------- ur_loader_interface.cpp - Level Zero Adapter ---------------===//
```

bader commented on sycl/plugins/unified_runtime/CMakeLists.txt:
```diff
@@ -93,18 +104,20 @@ add_sycl_plugin(unified_runtime
+  INCLUDE_DIRS
+  LIBRARIES
     UnifiedRuntime-Headers
+    #UnifiedRuntimeLoader
```
Why UnifiedRuntimeLoader is commented out?
Usually adding commented out code is considered to be a bad practice. Can we drop this line?

bader commented on sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_loader_interface.cpp:
```diff
@@ -0,0 +1,291 @@
+  return retVal;
+}
+
+/// @brief
+/// @param version
+/// @param pDdiTable
+/// @return
+UR_DLLEXPORT ur_result_t UR_APICALL urGetPlatformProcAddrTable(
```
```suggestion
UR_DLLEXPORT ur_result_t UR_APICALL urGetPlatformProcAddrTable(
```

bmyates replied:
@bader  I addressed all your review comments but now CI is failing due to a new unrelated issue :(

I think this change in sycl-tests broke something https://github.com/intel/llvm-test-suite/commit/a92d98e5936defd9d16dd7c5bab4023c24915b38

bader reviewed (approved):
Minor style comments.

smaslov-intel commented:
@bmyates : we expect a clean CI run, unless there is something known that @intel/llvm-gatekeepers would be aware of.
All failures need to be tracked (with either a PR to temporarily disabling failed tests, or a PR that fixes the test or compiler)
@intel/llvm-gatekeepers, is this a known issue:
> pi_die: Unknown parameter 4533 passed to hip_piKernelGetGroupInfo

bader commented:
> @bmyates : we expect a clean CI run, unless there is something known that @intel/llvm-gatekeepers would be aware of. All failures need to be tracked (with either a PR to temporarily disabling failed tests, or a PR that fixes the test or compiler) 

👍
That's right. In case of PR disabling failed test, we also expect an issue to re-enable the test back.

@intel/llvm-gatekeepers, is this a known issue:
> 
> > pi_die: Unknown parameter 4533 passed to hip_piKernelGetGroupInfo

I think @bmyates is right and this is a regression introduced by https://github.com/intel/llvm-test-suite/commit/a92d98e5936defd9d16dd7c5bab4023c24915b38. I see similar failure in https://github.com/intel/llvm/pull/8708 pre-commits.

bader commented:
@bmyates, could you fix, please?

```
/home/runner/work/llvm/llvm/src/sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero.cpp:1745:43: error: unused parameter 'device_flags' [-Werror,-Wunused-parameter]
ur_result_t urInit(ur_device_init_flags_t device_flags) {
                                          ^
/home/runner/work/llvm/llvm/src/sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero.cpp:1749:30: error: unused parameter 'pParams' [-Werror,-Wunused-parameter]
ur_result_t urTearDown(void *pParams) { return UR_RESULT_SUCCESS; }
                             ^
2 errors generated.
```
